python/lsst/cp/pipe/ptc.py

```diff
@@ -446,12 +413,42 @@ def runDataRef(self, dataRef, visitPairs):
 
             butler.put(linearizer, datasetType='Linearizer', dataId={'detector': detNum,
                        'detectorName': detName, 'calibDate': calibDate})
-
-        self.log.info(f"Writing PTC data to {dataRef.getUri(write=True)}")
+        rerunDir = list(dataRef.getButler().storage.repositoryCfgs.keys())[0]
+        self.log.info(f"Writing PTC data to {rerunDir} directory.")
         dataRef.put(datasetPtc, datasetType="photonTransferCurveDataset")
 
         return pipeBase.Struct(exitStatus=0)
 
+    def makePairs(self, dataRefList):
+        """Produce a list of flat pairs indexed by exposure time.
+
+        Parameters
+        ----------
+        dataReflist : `list` [`lsst.daf.peristence.ButlerDataRef`]
+            Data references for exposures for detectors to process.
+
+        Return
+        ------
+        sortedPairs : `dict`
+        """
+        sortedPairs = {}
+        for dataRef in dataRefList:
+            try:
+                tempFlat = dataRef.get("postISRCCD")
+            except RuntimeError:
+                self.log.warn(f"postISR exposure could not be retrieved. Ignoring flat.")
+                continue
+            expTime = tempFlat.getInfo().getVisitInfo().getExposureTime()
+            listAtExpTime = sortedPairs.setdefault(expTime, [])
+            if len(listAtExpTime) < 2:
+                listAtExpTime.append(tempFlat)
```

Comment:
Don't we want to sort on DATE-OBS or SEQNUM or something to prevent multiple flat ramps being paired incorrectly?  This also looks like it will toss any extra exposures, even when they could be used as well.

Reply:
I thought we were assuming for this task that the user would pass flat fields from a single PTC ramp. Are you suggesting that we should check that the DATE-OBS is the same for all exposures? Or do you have something else in mind? What if the flat ramp starts one night and continues to the next day, resulting in the images having different dates?  
The way we calculate the PTC now is by taking the difference between two flats at the same expTime to reduce PRNU. If for some reason there are more than two flats for the same expTime, we still need to pick two (the first two, in this code). Do you have something in mind for using possible extra exposures at the same expTime?

Reply:
I thought it would be general for any number of inputs.  The ramps that I'd looked at (from LATISS, though, so maybe different from how they're taking them on BOT) had four exposures at each exposure time.  More entries should provide better fits, as we're less dependent on any single pair.

Reply:
I do remember the LATIISS ramps with more than 2 exposures per exposure time, but I think that the PTC, as it's implemented now, uses the difference of only two. I'll add a message as you suggest.

Reply:
I think dropping exposures is a mistake.  This also isn't deterministic: for a set of equal exposure time exposures `1^2^3^4`, `1^3^2^4`, etc. will all give different outputs.  Even if you want to keep only one pair for each exposure time, you need to do some sort operation to ensure that order transposes all give the same result.

Reply:
VisitInfo does include the date of observation if that helps.

Reply:
The PTC task uses only two flats at a given exposure time. Should we require the user to make sure that this is the case? We could also look at the date of observation, as Tim suggests, but what if more than 2 flats have the same date of observation?

Reply:
How do you have two flats with the same date of observation and same detector? Isn't that impossible for a single instrument?

Reply:
I think this is the "date = %Y-%m-%d %H:%M:%S" vs "date = %Y-%m-%d" difference.  Sorting along the first (which I think is what `getVisitInfo().getDate()` returns) should sort this issue.

Reply:
I don't know, is it? If that is the case, how are you suggesting to use the date of observation to end up with only two flats at the same exposure time? (in case of Nflats with same expTime > 2).

Reply:
I just saw Chris's comment (after I posted mine, replying to Tim). Thanks!

Reply:
Yes, VisitInfo stores the full date, not just the day.

Reply:
This looks right to me.  If you can double check that it's deterministic (something like the example above where exposures are specified out of order but still return the in-order sorting), then I think this is done.